### PR TITLE
For #7249 Create a search widget

### LINKED ISSUE #7249: Create a search widget
### Why/User Benefit/User Problem
Android users are now used to engaging with the Google search widget in order to give the browser a signal that they desire to engage with a voice or text search directly from the home screen. For improved convenience Focus should help fulfill this use case.

### What / Requirements
As a user on the home screen I can initiate a text search by clicking on the search widget. 
As a user on the home screen I can initiate a voice search by clicking on the search widget. 
As a user I can customize the size of the search widget.
As a user I can add a widget to the home screen or remove it

### Acceptance Criteria (how do I know when I’m done?)
- If Focus is installed I can add a search widget from the home screen (a widget is listed under the Firefox entry in the widgets menu)
- If a widget is added to home screen I can resize it horizontally by long tapping it. As size reduces  only text search becomes possible
- If a widget is added to home screen I can remove it
- If Focus is uninstalled the widget gets removed from the home screen
- The widget has a focus logo, some text that signals that users can text search and a microphone icon that signals that users can voice search
- If a user engages with text search, Focus opens with focus on the search bar and keyboard opened
- If a user engages with voice search, the Google voice recognition capability opens to start recording the voice. After a timeout either the Google voice recognition capability prompts the user to say something (no voice was detected), or Focus opens search results for the recognized string of text (voice was detected) using the default search provider
- Ensure widget design is extensible to allow for future addition of QR code scanning
- Event Telemetry identifies that the application was launched using the search widget (text or voice search can be segmented)
- Environment telemetry identifies that the search widget is installed

### PR DESCRIPTION
**Pull request by iorgamgabriel**
### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR:
1. click on `Show All Checks`,
2. click `Details` next to `build-focus-debug` or `build-klar-debug` for changes targeting Klar,
2. click `View task in Taskcluster`,
3. click the `Artifacts` row,
4. click to download any of the apks listed here which use an appropriate name for each CPU architecture.

### GitHub Automation
Fixes #7249

### CODE REVIEW

File: app/src/main/java/org/mozilla/gecko/search/SearchWidgetProvider.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import mozilla.components.feature.search.widget.SearchWidgetConfig
+import mozilla.components.support.utils.PendingIntentUtils
+import org.mozilla.focus.R
+import org.mozilla.focus.activity.IntentReceiverActivity
+import org.mozilla.focus.ext.components
+import org.mozilla.focus.searchwidget.VoiceSearchActivity
+
+class SearchWidgetProvider : AppSearchWidgetProvider() {
+
+    // Implementation note:
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/gecko/search/SearchWidgetProvider.kt:21]: This note does not apply to Focus.

File: app/src/main/java/org/mozilla/gecko/search/SearchWidgetProvider.kt [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    override val config: SearchWidgetConfig =
+        SearchWidgetConfig(
+            searchWidgetIconResource = R.drawable.ic_splash_screen,
+            searchWidgetMicrophoneResource = R.drawable.mozac_ic_microphone,
+            appName = R.string.app_name
+        )
+
+    override fun createTextSearchIntent(context: Context): PendingIntent {
+        return Intent(context, IntentReceiverActivity::class.java)
+            .let { intent ->
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/gecko/search/SearchWidgetProvider.kt:42]: nit: The `let` usage and formatting caught my attention. The construction of the PendingIntent seems a bit confusing.
Can we split this in two steps: first constructing the Intent with maybe using `apply` and finally building and returning the PendingIntent?

File: app/src/main/java/org/mozilla/focus/FocusApplication.kt [hunk 3]
```diff
@@ -87,6 +88,8 @@ open class FocusApplication : LocaleAwareApplication(), CoroutineScope {
             components.appStartReasonProvider.registerInAppOnCreate(this)
             components.startupActivityLog.registerInAppOnCreate(this)
 
+            Metrics.searchWidgetInstalled.set(settings.searchWidgetInstalled)
```
Comment (Mugurell) [inline app/src/main/java/org/mozilla/focus/FocusApplication.kt:91]: Can this be moved in [`collectPrefMetrics`](https://github.com/mozilla-mobile/focus-android/blob/1aca30c95aad8b9c15b88375a379770daf8c3d66/app/src/main/java/org/mozilla/focus/telemetry/GleanMetricsService.kt#L93)?

**Review by Mugurell — APPROVED** [review]
Some small nits but this LGTM.
r+ for the code.

File: app/src/main/java/org/mozilla/gecko/search/SearchWidgetProvider.kt (hunk 1 above)
Comment (Mugurell) [reply app/src/main/java/org/mozilla/gecko/search/SearchWidgetProvider.kt:21]: Also about this. We don't need the gecko package for Focus. That was a requirement just for supporting the Fennec -> Fenix migration.

Comment (iorgamgabriel) [reply app/src/main/java/org/mozilla/gecko/search/SearchWidgetProvider.kt]: Done

File: app/src/main/java/org/mozilla/focus/FocusApplication.kt (hunk 3 above)
Comment (iorgamgabriel) [reply app/src/main/java/org/mozilla/focus/FocusApplication.kt]: Done

**Comment by iorgamgabriel** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?

- What are the most common types of start up?
- For users in the 95th percentile of slow start ups – we assume the most likely to leave the app – what are the most common types of start ups?

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:

- We want to know the most common types of start up so that we can invest perf resources into only the most common types

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

None: we need to understand user behaviour in the wild

4) Can current instrumentation answer these questions?

An existing probe `app_opened_all_startup`, tries to measure this but we found one thing wrong with it so I doubt its accuracy. So no.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

<table>
  <tr>
    <td>perf.startup.startup_type</td>
    <td>2 - interaction</td>
    <td>#7079</td>
  </tr>
</table>

6) How long will this data be collected?  Choose one of the following:

* This technical data is intended to be collected perpetually.

7) What populations will you measure?

* Which release channels?

All

* Which countries?

All

* Which locales?

All

* Any other filters?  Please describe in detail below.

N/A

8) If this data collection is default on, what is the opt-out mechanism for users?

Telemetry opt out in app

9) Please provide a general description of how you will analyze this data.

- Compare different labels in GLAM
- Write percentile query but filter for 95th percentile slow starts (probable heuristic instead: low-end devices)

10) Where do you intend to share the results of your analysis?

Internally: GLAM, sql.tmo, gdocs

11) Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection? If so:

No

**Comment by rocketsroger** [comment]
Please make sure you answer question 5 for the data review request.  Thanks

**Comment by rocketsroger** [comment]
> # Request for data collection review form
> 
> All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.
> 
>     1. What questions will you answer with this data?
> 
> 
>     * search_widget_installed - If the user has the search widget added on the home screen
> 
>     * new_tab_button - When the user has pressed anywhere from the Focus logo until the start of the microphone icon, opening a new tab search screen.
> 
>     * voice_button - When the user pressed the microphone icon, opening a new voice search screen.
> 
> 
>     2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements?
> 
> 
>     * We are adding new features in history to make users' interactions more efficient.
> 
> 
>     3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
> 
> 
>     * There are no other alternatives.
> 
> 
>     4. Can current instrumentation answer these questions?
> 
> 
>     * No.
> 
> 
>     5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki.    
>        Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.
> 
> 
> Measurement Description 	Data Collection Category 	Tracking Bug #
> Search Widget Installed: If the user has the search widget added on the home screen. 	Category 2 - interaction data 	#7249
> Search Widget New Tab Button: When the user has pressed anywhere from the Focus logo until the start of the microphone icon, opening a new tab search screen. 	Category 2 - interaction data 	#7249
> Search Widget Voice Button: When the user pressed the microphone icon, opening a new voice search screen. 	Category 2 - interaction data 	#7249
> 
>     6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
> 
> 
>     * https://dictionary.telemetry.mozilla.org/apps/focus_android
> 
> 
>     7. How long will this data be collected? Choose one of the following:
> 
> 
>     * Up until version 119, with the option to renew at that point.
> 
> 
>     8. What populations will you measure?
> 
> 
>     * All channels, all locales, all countries
> 
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
> 
> 
>     * Default Glean SDK opt-out mechanism.
> 
> 
>     10. Please provide a general description of how you will analyze this data.
> 
> 
>     * GlAM.
> 
> 
>     11. Where do you intend to share the results of your analysis?
> 
> 
>     * Only on Glean, Glam, and with mobile teams.
> 
> 
>     12. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
> 
> 
>     * No third-party tools.

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way? 

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/apps/fenix )

2) Is there a control mechanism that allows the user to turn the data collection on and off? 

Yes, through the "Send Usage Data" preference in the application settings

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

N/A, collection set to end or be renewed by version 119

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 2, Interaction data

5) Is the data collection request for default-on or default-off?

default-on

6) Does the instrumentation include the addition of **any *new* identifiers**?

No

7) Is the data collection covered by the existing Firefox privacy notice? 

Yes

8) Does the data collection use a third-party collection tool? 

No

## Result

data-review+

**Review by Mugurell — APPROVED** [review]
Looks and functions nicely 👍 ✨

File: app/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -131,6 +131,10 @@
         <activity android:name=".activity.EraseAndOpenShortcutActivity"
             android:theme="@android:style/Theme.Translucent" />
 
+        <activity
+            android:name=".searchwidget.VoiceSearchActivity"
+            android:theme="@style/Theme.AppCompat.Translucent" />
+
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply app/src/main/AndroidManifest.xml:137]: Done

File: app/src/main/java/org/mozilla/focus/activity/IntentReceiverActivity.kt [hunk 5]
```diff
@@ -58,7 +63,15 @@ class IntentReceiverActivity : Activity() {
...
+        intent.putExtra(SEARCH_WIDGET, intent.getBooleanExtra(SEARCH_WIDGET, false))
+        intent.putExtra(
+            BaseVoiceSearchActivity.SPEECH_PROCESSING,
+            intent.getStringExtra(BaseVoiceSearchActivity.SPEECH_PROCESSING)
+        )
         startActivity(intent)
     }
+
+    companion object {
+        const val SEARCH_WIDGET = "search_widget"
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply app/src/main/java/org/mozilla/focus/activity/IntentReceiverActivity.kt]: Done

File: app/src/main/java/org/mozilla/focus/activity/IntentReceiverActivity.kt [hunk 6]
```diff
@@ -25,14 +28,15 @@ class IntentReceiverActivity : Activity() {
         super.onCreate(savedInstanceState)
 
         val intent = intent.sanitize().toSafeIntent()
-
+        if (intent.getBooleanExtra(SEARCH_WIDGET, false)) {
+            SearchWidget.newTabButton.record(NoExtras())
+        }
```
[the remark this answers is not in the document]
Comment (iorgamgabriel) [reply app/src/main/java/org/mozilla/focus/activity/IntentReceiverActivity.kt]: Done

**Pull request merged**